### DISCUSSION TITLE
Promethues/Grafana Setup

### DISCUSSION THREAD

**Discussion by jamannmtgov** [netbox-community/netbox#12309] [Q&A]
Curious if anyone has a more detailed setup instructions/examples to setup Prometheus and Grafana for Netbox? I am a completely new to Prometheus and Grafana but I was able to follow the Monitoring instructions for the docker image but not sure what to do in Grafana? Any help would be appreciated, thanks!

**Comment by candlerb** [comment]
> I was able to follow the Monitoring instructions for the docker image

Which monitoring instructions are you referring to, and for which docker image?  If you can say what you've already read and built, it may be easier to fill in the gaps.

Are you trying to monitor Netbox itself using Prometheus and Grafana? Or are you trying to use Netbox as an inventory to tell Prometheus and Grafana what to monitor? I'm going to assume the latter.

Before you can use Grafana, you need to be storing metrics data in Prometheus. If you're collecting information from Linux servers using node_exporter, there are a number of dashboards you can use ([1860](https://grafana.com/grafana/dashboards/1860-node-exporter-full/) is one of them). If you're collecting SNMP MIBs from network devices then you may have to build your own dashboards, although I created 12489 and 12492 as examples using the if_mib.

As regards Netbox integration: the way I have it set up is Prometheus to use Netbox as its database of targets to scrape (service discovery) using [netbox-plugin-prometheus-sd](https://github.com/FlxPeters/netbox-plugin-prometheus-sd).  I use tags to identify which devices are to be scraped using node_exporter and which using snmp_exporter, and a custom field to identify which module to use for snmp_exporter. You need to use Prometheus rewriting rules to change the Netbox labels into things you can use (e.g. map the "primary address" to the `__address__` to scrape)

You can find some example configs in the lab exercises on days 4 and 5 of [this workshop](https://nsrc.org/workshops/2022/rwnog/nmm/agenda.html), and in the "supplementary materials" at the end. These are the most relevant to your question:
* [Grafana as dashboard front-end to Prometheus](https://nsrc.org/workshops/2022/rwnog/nmm/netmgmt/en/prometheus/ex-grafana.html)
* [Integrating Prometheus with Netbox](https://nsrc.org/workshops/2022/rwnog/nmm/netmgmt/en/prometheus/ex-integration.html)

but they do assume that previous exercises have been done to install and configure prometheus, node_exporter, snmp_exporter etc.

Note: this integration is using prometheus as a "source of truth" to *configure* Prometheus, i.e. tell it what to monitor. If you want to see the *current* state of devices, e.g. traffic levels and up/down status, you'd do that in the Grafana or Prometheus web interface. You won't be importing data *into* Netbox.

For Prometheus as a network monitoring tool, there are a number of current gaps in capability which are being worked on, see this document: https://docs.google.com/document/d/1oEpjiWfTHF352NCAOGolwij3EIkrprCkdQmaQMpjg4M/edit
